// zotero-api-client/lib/main.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime-corejs3/helpers/defineProperty');
var _endsWithInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/ends-with');
var _startsWithInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/starts-with');
var _includesInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/includes');
var _Promise = require('@babel/runtime-corejs3/core-js-stable/promise');
var _trimInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/trim');
var SparkMD5 = require('spark-md5');
var _parseInt = require('@babel/runtime-corejs3/core-js-stable/parse-int');
var _matchAllInstanceProperty = require('@babel/runtime-corejs3/core-js/instance/match-all');
var _reduceInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/reduce');

function ownKeys$2(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread$2(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys$2(Object(t), true).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys$2(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
const parseIntHeaders = (headers, headerName) => {
  const value = headers && headers.get(headerName) || null;
  return value === null ? null : _parseInt(value, 10);
};

/**
 * @class Represents a generic Zotero API response. Usually a specialised variant inheriting from
 * this class is returned when doing an API request
 * @memberof module:zotero-api-client
 * @inner
 */
class ApiResponse {
  constructor(data, options, response) {
    this.raw = data;
    this.options = options;
    this.response = response;
  }

  /**
   * Name of the class, useful to determine instance of which specialised class
    has been returned
   * @return {string} name of the class
   */
  getResponseType() {
    return 'ApiResponse';
  }

  /**
   * Content of the response. Specialised classes provide extracted data depending on context.
   * @return {object}
   */
  getData() {
    return this.raw;
  }

  /**
   * Links available in the response. Specialised classes provide extracted links depending on context.
   * @return {object}
   */
  getLinks() {
    if ('links' in this.raw) {
      return this.raw.links;
    }
    return null;
  }

  /**
   * Meta data available in the response. Specialised classes provide extracted meta data depending on context.
   * @return {object}
   */
  getMeta() {
    if ('meta' in this.raw) {
      return this.raw.meta;
    }
    return null;
  }

  /**
   * Value of the "Last-Modified-Version" header in response if present. Specialised classes provide
    version depending on context
   * @return {?number} Version of the content in response
   */
  getVersion() {
    return parseIntHeaders(this.response.headers, 'Last-Modified-Version');
  }
}
class SchemaResponse extends ApiResponse {
  /**
   * @see {@link module:zotero-api-client~ApiResponse#getResponseType}
   */
  getResponseType() {
    return 'SchemaResponse';
  }

  /**
   * @returns {number} Version of the schema
   */
  getVersion() {
    return this.raw.version;
  }
  getMeta() {
    return null;
  }
}

/**
 * @class Represents a response to a GET request containing a single entity
 * @extends ApiResponse
 * @memberof module:zotero-api-client
 * @inner
 */
class SingleReadResponse extends ApiResponse {
  /**
   * @see {@link module:zotero-api-client~ApiResponse#getResponseType}
   */
  getResponseType() {
    return 'SingleReadResponse';
  }

  /**
   * @return {Object} entity returned in this response
   */
  getData() {
    return this.raw && 'data' in this.raw ? this.raw.data : this.raw;
  }
}

/**
 * @class represents a response to a GET request containing multiple entities
 * @extends ApiResponse
 * @memberof module:zotero-api-client
 * @inner
 */
class MultiReadResponse extends ApiResponse {
  /**
   * @see {@link module:zotero-api-client~ApiResponse#getResponseType}
   */
  getResponseType() {
    return 'MultiReadResponse';
  }

  /**
   * @return {Array} a list of entities returned in this response
   */
  getData() {
    return this.raw.map(r => 'data' in r ? r.data : 'tag' in r ? {
      tag: r.tag
    } : r);
  }

  /**
   * @return {Array} a list of links, indexes of the array match indexes of entities in {@link
  module:zotero-api-client~MultiReadResponse#getData}
   */
  getLinks() {
    return this.raw.map(r => 'links' in r && r.links || null);
  }

  /**
   * @return {Array} a list of meta-data, indexes of the array match indexes of entities in {@link
  module:zotero-api-client~MultiReadResponse#getData}
   */
  getMeta() {
    return this.raw.map(r => 'meta' in r && r.meta || null);
  }

  /**
   * @return {?number} Total number of results
   */
  getTotalResults() {
    return parseIntHeaders(this.response.headers, 'Total-Results');
  }

  /**
   * @return {object} Parsed content of "Link" header as an object where value of "rel" is a key and
    the URL is the value. For paginated responses contain URLs for "first", "next", "prev" and "last".
   */
  getRelLinks() {
    var _this$response$header, _context;
    const links = (_this$response$header = this.response.headers.get('link')) !== null && _this$response$header !== void 0 ? _this$response$header : '';
    const matches = Array.from(_matchAllInstanceProperty(links).call(links, /<(.*?)>;\s+rel="(.*?)"/ig));
    return _reduceInstanceProperty(_context = Array.from(matches)).call(_context, (acc, match) => {
      const url = match[1];
      const rel = match[2];
      acc[rel] = url;
      return acc;
    }, {});
  }
}

/**
 * @class Represents a response to a PUT or PATCH request
 * @extends ApiResponse
 * @memberof module:zotero-api-client
 * @inner
 */
class SingleWriteResponse extends ApiResponse {
  /**
   * @see {@link module:zotero-api-client~ApiResponse#getResponseType}
   */
  getResponseType() {
    return 'SingleWriteResponse';
  }

  /**
   * @return {Object} For put requests, this represents a complete, updated object.
   *                  For patch requests, this represents only updated fields of the updated object.
   */
  getData() {
    return _objectSpread$2(_objectSpread$2({}, this.options.body), {}, {
      version: this.getVersion()
    });
  }
}

/**
 * @class Represents a response to a POST request
 * @extends ApiResponse
 * @memberof module:zotero-api-client
 * @inner
 */
class MultiWriteResponse extends ApiResponse {
  /**
   * @see {@link module:zotero-api-client~ApiResponse#getResponseType}
   */
  getResponseType() {
    return 'MultiWriteResponse';
  }

  /**
   * @return {Boolean} Indicates whether all write operations were successful
   */
  isSuccess() {
    return Object.keys(this.raw.failed).length === 0;
  }

  /**
   * Returns all entities POSTed in an array. Entities that have been written successfully
   * are returned updated, other entities are returned unchanged. It is advised to verify
   * if the request was entirely successful (see isSuccess and getErrors) before using this method.
   * @return {Array} A modified list of all entities posted.
   */
  getData() {
    return this.options.body.map((item, index) => {
      index = index.toString();
      if (index in this.raw.success) {
        const remoteItem = this.raw.successful && this.raw.successful[index] || {};
        return _objectSpread$2(_objectSpread$2(_objectSpread$2({}, item), remoteItem.data), {}, {
          key: this.raw.success[index],
          version: this.getVersion()
        });
      } else {
        return item;
      }
    });
  }

  /**
   * @see {@link module:zotero-api-client~ApiResponse#getLinks}
   */
  getLinks() {
    return this.options.body.map((_, index) => {
      if ("successful" in this.raw) {
        const entry = this.raw.successful[index.toString()];
        if (entry) {
          return entry.links || null;
        }
      }
      return null;
    });
  }

  /**
   * @see {@link module:zotero-api-client~ApiResponse#getMeta}
   */
  getMeta() {
    return this.options.body.map((_, index) => {
      if ("successful" in this.raw) {
        const entry = this.raw.successful[index.toString()];
        if (entry) {
          return entry.meta || null;
        }
      }
      return null;
    });
  }

  /**
   * Returns all errors that have occurred.
   * @return {Object} Errors object where keys are indexes of the array of the original request and values are the errors occurred.
   */
  getErrors() {
    const errors = {};
    for (let i of Object.keys(this.raw.failed)) {
      errors[_parseInt(i, 10)] = this.raw.failed[i];
    }
    return errors;
  }

  /**
   * Allows getting an updated entity based on its key, otherwise identical to getEntityByIndex
   * @param  {String} key
   * @return {Object}
   * @throws {Error} If key is not present in the request
   * @see {@link module:zotero-api-client~MultiWriteResponse#getEntityByIndex}
   */
  getEntityByKey(key) {
    let index = this.options.body.findIndex(entity => {
      return entity.key === key;
    });
    if (index > -1) {
      return this.getEntityByIndex(index);
    }
    throw new Error("Key ".concat(key, " is not present in the request"));
  }

  /**
   * Allows getting an updated entity based on its index in the original request
   * @param  {Number|String} index
   * @return {Object}
   * @throws {Error} If index is not present in the original request
   * @throws {Error} If error occurred in the POST for selected entity. Error message will contain the reason for failure.
   */
  getEntityByIndex(index) {
    if (typeof index === 'string') {
      index = _parseInt(index, 10);
    }
    if (index.toString() in this.raw.success) {
      const remoteItem = this.raw.successful && this.raw.successful[index.toString()] || {};
      return _objectSpread$2(_objectSpread$2(_objectSpread$2({}, this.options.body[index]), remoteItem.data), {}, {
        key: this.raw.success[index],
        version: this.getVersion()
      });
    }
    if (index.toString() in this.raw.unchanged) {
      return this.options.body[index];
    }
    if (index.toString() in this.raw.failed) {
      throw new Error("".concat(this.raw.failed[index.toString()].code, ": ").concat(this.raw.failed[index.toString()].message));
    }
    throw new Error("Index ".concat(index, " is not present in the response"));
  }
}

/**
 * @class Represents a response to a DELETE request
 * @extends ApiResponse
 * @memberof module:zotero-api-client
 * @inner
 */
class DeleteResponse extends ApiResponse {
  /**
   * @see {@link module:zotero-api-client~ApiResponse#getResponseType}
   */
  getResponseType() {
    return 'DeleteResponse';
  }
}

/**
 * @class Represents a response to a file upload request
 * @extends ApiResponse
 * @property {Object} authResponse     - Response object for stage 1 (upload authorisation) request
 * @property {Object} response 	       - alias for "authResponse"
 * @property {Object} uploadResponse   - Response object for stage 2 (file upload) request
 * @property {Object} registerResponse - Response object for stage 3 (upload registration) request
 * @memberof module:zotero-api-client
 * @inner
 */
class FileUploadResponse extends ApiResponse {
  constructor(data, options, authResponse, uploadResponse, registerResponse) {
    super(data, options, authResponse);
    this.uploadResponse = uploadResponse;
    this.registerResponse = registerResponse;
  }

  /**
   * @see {@link module:zotero-api-client~ApiResponse#getResponseType}
   */
  getResponseType() {
    return 'FileUploadResponse';
  }

  /**
   * @see {@link module:zotero-api-client~ApiResponse#getVersion}
   */
  getVersion() {
    var _ref, _parseIntHeaders, _this$registerRespons, _this$uploadResponse;
    // full upload will have the latest version in the final register response,
    // partial upload could have the latest version in the upload response
    // (because that goes through API), however, currently that's not the case.
    // If a file existed before, and currently for partial uploads, the latest version
    // will be in obtained from the initial response
    return (_ref = (_parseIntHeaders = parseIntHeaders((_this$registerRespons = this.registerResponse) === null || _this$registerRespons === void 0 ? void 0 : _this$registerRespons.headers, 'Last-Modified-Version')) !== null && _parseIntHeaders !== void 0 ? _parseIntHeaders : parseIntHeaders((_this$uploadResponse = this.uploadResponse) === null || _this$uploadResponse === void 0 ? void 0 : _this$uploadResponse.headers, 'Last-Modified-Version')) !== null && _ref !== void 0 ? _ref : parseIntHeaders(this.response.headers, 'Last-Modified-Version');
  }
}

/**
 * @class Represents a response to a file download request
 * @extends ApiResponse
 * @memberof module:zotero-api-client
 * @inner
 */
class FileDownloadResponse extends ApiResponse {
  /**
   * @see {@link module:zotero-api-client~ApiResponse#getResponseType}
   */
  getResponseType() {
    return 'FileDownloadResponse';
  }
}

/**
 * @class Represents a response containing temporary url for file download
 * @extends ApiResponse
 * @memberof module:zotero-api-client
 * @inner
 */
class FileUrlResponse extends ApiResponse {
  /**
   * @see {@link module:zotero-api-client~ApiResponse#getResponseType}
   */
  getResponseType() {
    return 'FileUrlResponse';
  }
}

/**
 * @class Represents a raw response, e.g. to data requests with format other than JSON
 * @extends ApiResponse
 * @memberof module:zotero-api-client
 * @inner
 */
class RawApiResponse extends ApiResponse {
  constructor(rawResponse, options) {
    super(rawResponse, options, rawResponse);
  }

  /**
   * @see {@link module:zotero-api-client~ApiResponse#getResponseType}
   */
  getResponseType() {
    return 'RawApiResponse';
  }
}

/**
 * @class Represents a response for pretended request, mostly for debug purposes. See {@link module:zotero-api-client.api~pretend}
 * @extends ApiResponse
 * @memberof module:zotero-api-client
 * @inner
 */
class PretendResponse extends ApiResponse {
  /**
   * @see {@link module:zotero-api-client~ApiResponse#getResponseType}
   */
  getResponseType() {
    return 'PretendResponse';
  }

  /**
   * @return {Object} For pretended request version will always be null.
   */
  getVersion() {
    return null;
  }
}

/**
 * @class Represents an error response from the api
 * @extends Error
 * @property {Object} response - Response object for the request, with untouched body
 * @property {String} message  - What error occurred, usually contains response code and status
 * @property {String} reason   - More detailed reason for the failure, if provided by the API
 * @property {Object} options  - Configuration object used for this request
 * @memberof module:zotero-api-client
 * @inner
 */
class ErrorResponse extends Error {
  constructor(message, reason, response, options) {
    super(message);
    this.response = response;
    this.reason = reason;
    this.message = message;
    this.options = options;
  }

  /**
   * Value of the "Last-Modified-Version" header in response if present. This is generally only available if the server responded with 412 due to a version mismatch.
   * @return {?number} Version of the content in response
   */
  getVersion() {
    return parseIntHeaders(this.response.headers, 'Last-Modified-Version');
  }

  /**
   * @see {@link module:zotero-api-client~ApiResponse#getResponseType}
   */
  getResponseType() {
    return 'ErrorResponse';
  }
}

function ownKeys$1(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread$1(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys$1(Object(t), true).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys$1(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
const acceptedHeaderNames = {
  authorization: 'Authorization',
  contentType: 'Content-Type',
  ifMatch: 'If-Match',
  ifModifiedSinceVersion: 'If-Modified-Since-Version',
  ifNoneMatch: 'If-None-Match',
  ifUnmodifiedSinceVersion: 'If-Unmodified-Since-Version',
  zoteroApiKey: 'Zotero-API-Key',
  zoteroWriteToken: 'Zotero-Write-Token',
  zoteroSchemaVersion: 'Zotero-Schema-Version'
};
const queryParamsWithArraySupport = ["tag", "itemTag"];
const queryParamNames = ['annotationType', 'collectionKey', 'content', 'direction', 'format', 'include', 'includeTrashed', 'itemKey', 'itemQ', 'itemQMode', 'itemTag', 'itemType', 'limit', 'linkMode', 'linkwrap', 'locale', 'q', 'qmode', 'searchKey', 'since', 'sort', 'start', 'style', 'tag'];
const filePatchQueryParamNames = ['algorithm', 'upload'];
const fetchParamNames = ['body', 'cache', 'credentials', 'integrity', 'keepalive', 'method', 'mode', 'priority', 'redirect', 'referrer', 'referrerPolicy', 'signal'];
const resourcesSpecs = [
//name in resource, name in the url (usually the same, but there are exceptions)
{
  'name': 'library',
  urlPart: 'library',
  isKeyResource: true
}, {
  'name': 'collections',
  urlPart: 'collections',
  isKeyResource: true
}, {
  'name': 'publications',
  urlPart: 'publications',
  isKeyResource: false
}, {
  'name': 'items',
  urlPart: 'items',
  isKeyResource: true
}, {
  'name': 'searches',
  urlPart: 'searches',
  isKeyResource: true
}, {
  'name': 'top',
  urlPart: 'top',
  isKeyResource: false
}, {
  'name': 'trash',
  urlPart: 'trash',
  isKeyResource: false
}, {
  'name': 'tags',
  urlPart: 'tags',
  isKeyResource: true
}, {
  'name': 'children',
  urlPart: 'children',
  isKeyResource: false
}, {
  'name': 'groups',
  urlPart: 'groups',
  isKeyResource: false
}, {
  'name': 'subcollections',
  urlPart: 'collections',
  isKeyResource: false
}, {
  'name': 'itemTypes',
  urlPart: 'itemTypes',
  isKeyResource: false
}, {
  'name': 'itemFields',
  urlPart: 'itemFields',
  isKeyResource: false
}, {
  'name': 'schema',
  urlPart: 'schema',
  isKeyResource: false
}, {
  'name': 'creatorFields',
  urlPart: 'creatorFields',
  isKeyResource: false
}, {
  'name': 'itemTypeFields',
  urlPart: 'itemTypeFields',
  isKeyResource: false
}, {
  'name': 'itemTypeCreatorTypes',
  urlPart: 'itemTypeCreatorTypes',
  isKeyResource: false
}, {
  'name': 'template',
  urlPart: 'items/new',
  isKeyResource: false
}, {
  'name': 'file',
  urlPart: 'file',
  isKeyResource: false
}, {
  'name': 'fileUrl',
  urlPart: 'file/view/url',
  isKeyResource: false
}, {
  'name': 'settings',
  urlPart: 'settings',
  isKeyResource: true
}, {
  'name': 'deleted',
  'urlPart': 'deleted',
  isKeyResource: false
}, {
  'name': 'verifyKeyAccess',
  urlPart: 'keys/current',
  isKeyResource: false
}];
const defaults = {
  apiAuthorityPart: 'api.zotero.org',
  apiPath: '',
  apiScheme: 'https',
  cache: 'default',
  credentials: 'omit',
  format: 'json',
  method: 'get',
  mode: 'cors',
  pretend: false,
  redirect: 'follow',
  resource: {},
  retry: 0,
  retryDelay: null,
  uploadRegisterOnly: null
};
const makeUrlPath = resource => {
  let path = [];
  for (let resourcesSpec of resourcesSpecs) {
    let resourceValue = resource[resourcesSpec.name];
    if (resourcesSpec.name in resource) {
      if (resourcesSpec.name === 'library') {
        if (resourceValue.charAt(0) === 'u') {
          path.push('users', resourceValue.slice(1));
        } else if (resourceValue.charAt(0) === 'g') {
          path.push('groups', resourceValue.slice(1));
        }
      } else {
        path.push(resourcesSpec.urlPart);
        if (resourcesSpec.isKeyResource && resource[resourcesSpec.name]) {
          path.push(resourceValue);
        }
      }
    }
  }
  return path.join('/');
};
const makeUrlQuery = (options, paramNames) => {
  let params = [];
  for (let name of paramNames) {
    if (options[name]) {
      if (_includesInstanceProperty(queryParamsWithArraySupport).call(queryParamsWithArraySupport, name) && Array.isArray(options[name])) {
        params.push(...options[name].map(k => "".concat(name, "=").concat(encodeURIComponent(k))));
      } else {
        params.push("".concat(name, "=").concat(encodeURIComponent(options[name])));
      }
    }
  }
  return params.length ? '?' + params.join('&') : '';
};
const makeHeaders = (options, headerNames) => {
  let headers = {};
  for (let header of Object.keys(headerNames)) {
    if (header in options) {
      headers[headerNames[header]] = options[header];
    }
  }
  return headers;
};
const hasDefinedKey = (object, key) => {
  return key in object && object[key] !== null && typeof object[key] !== 'undefined';
};
const throwErrorResponse = async (rawResponse, options, requestDesc) => {
  let clonedRawResponse = rawResponse.clone();
  let reason = await clonedRawResponse.text();
  throw new ErrorResponse("".concat(requestDesc).concat(rawResponse.status, ": ").concat(rawResponse.statusText), reason, rawResponse, options);
};
const isTransientFailure = response => response.status === 408 || response.status >= 500;
const sleep = seconds => {
  return new _Promise(resolve => {
    setTimeout(() => {
      resolve();
    }, seconds * 1000);
  });
};

/**
 * Executes request and returns a response. Not meant to be called directly, instead use {@link
   module:zotero-api-client~api}.
 * @param {Object} config - Configuration object
 * @param {String} config.apiScheme - Scheme part of the API URL
 * @param {String} config.apiAuthorityPart - Authority part of the API URL
 * @param {String} config.apiPath - Path part of the API URL
 * @param {String} config.authorization - 'Authorization' header
 * @param {String} config.zoteroWriteToken - 'Zotero-Write-Token' header
 * @param {String} config.ifModifiedSinceVersion - 'If-Modified-Since-Version' header
 * @param {String} config.ifUnmodifiedSinceVersion - 'If-Unmodified-Since-Version' header
 * @param {String} config.contentType - 'Content-Type' header
 * @param {String} config.collectionKey - 'collectionKey' query argument
 * @param {String} config.content - 'content' query argument
 * @param {String} config.direction - 'direction' query argument
 * @param {String} config.format - 'format' query argument
 * @param {String} config.include - 'include' query argument
 * @param {String} config.includeTrashed - 'includeTrashed' query argument
 * @param {String} config.itemKey - 'itemKey' query argument
 * @param {String} config.itemQ - 'itemQ' query argument
 * @param {String} config.itemQMode - 'itemQMode' query argument
 * @param {String|String[]} config.itemTag - 'itemTag' query argument
 * @param {String} config.itemType - 'itemType' query argument
 * @param {Number} config.limit - 'limit' query argument
 * @param {String} config.linkMode - 'linkMode' query argument
 * @param {String} config.linkwrap - 'linkwrap' query argument
 * @param {String} config.locale - 'locale' query argument
 * @param {String} config.q - 'q' query argument
 * @param {String} config.qmode - 'qmode' query argument
 * @param {String} config.searchKey - 'searchKey' query argument
 * @param {Number} config.since - 'since' query argument
 * @param {String} config.sort - 'sort' query argument
 * @param {Number} config.start - 'start' query argument
 * @param {String} config.style - 'style' query argument
 * @param {String|String[]} config.tag - 'tag' query argument
 * @param {Boolean} config.pretend - triggers pretend mode where fetch request is prepared and returned without execution
 * @param {String} config.resource.top - use 'top' resource
 * @param {String} config.resource.trash - use 'trash' resource
 * @param {String} config.resource.children - use 'children' resource
 * @param {String} config.resource.groups - use 'groups' resource
 * @param {String} config.resource.itemTypes - use 'itemTypes' resource
 * @param {String} config.resource.itemFields - use 'itemFields' resource
 * @param {String} config.resource.creatorFields - use 'creatorFields' resource
 * @param {String} config.resource.itemTypeFields - use 'itemTypeFields' resource
 * @param {String} config.resource.itemTypeCreatorTypes - use 'itemTypeCreatorTypes' resource
 * @param {String} config.resource.library - use 'library' resource
 * @param {String} config.resource.collections - use 'collections' resource
 * @param {String} config.resource.items - use 'items' resource
 * @param {String} config.resource.searches - use 'searches' resource
 * @param {String} config.resource.tags - use 'tags' resource
 * @param {String} config.resource.template - use 'template' resource
 * @param {String} config.method - forwarded to fetch()
 * @param {String} config.body - forwarded to fetch()
 * @param {String} config.mode - forwarded to fetch()
 * @param {String} config.cache - forwarded to fetch()
 * @param {String} config.credentials - forwarded to fetch()
 * @param {Boolean} config.uploadRegisterOnly - this file upload should only perform stage 1
 * @param {Number} config.retry - retry this many times after transient error
 * @param {Number} config.retryDelay - wait this many seconds before retry. If not set an exponential backoff algorithm will be used
 *
 * @return {Promise} Returns a Promise that will eventually return a response object
 * @throws {Error} If options specify impossible configuration
 * @throws {ErrorResponse} If API responds with a non-ok response
 * @memberof module:zotero-api-client
 * @inner
 */
const request = async config => {
  var _context;
  let response;
  if ('response' in config && config.response) {
    return config;
  }
  const options = _objectSpread$1(_objectSpread$1({}, defaults), config);
  if (hasDefinedKey(options, 'body') && (hasDefinedKey(options, 'file') || hasDefinedKey(options, 'oldFile'))) {
    throw new Error('Cannot use both "file" and "body" in a single request.');
  }
  if (_includesInstanceProperty(_context = ['POST', 'PUT', 'PATCH']).call(_context, options.method.toUpperCase()) && !('contentType' in config)) {
    options.contentType = 'application/json';
  }
  if (hasDefinedKey(options, 'filePatch')) {
    // for partial file upload api uses patch(), however first request (file authorisation still uses POST
    options.method = 'POST';
  }
  const headers = makeHeaders(options, acceptedHeaderNames);
  const path = makeUrlPath(options.resource);
  const query = makeUrlQuery(options, queryParamNames);
  const url = "".concat(options.apiScheme, "://").concat(options.apiAuthorityPart, "/").concat(options.apiPath).concat(path).concat(query);
  const fetchConfig = {};
  for (let param of fetchParamNames) {
    if (param === 'body' && hasDefinedKey(options, param)) {
      fetchConfig[param] = JSON.stringify(options[param]);
    } else {
      fetchConfig[param] = options[param];
    }
  }
  let fileUploadData = {};
  // process the request for file upload authorisation request
  if ((hasDefinedKey(options, 'filePatch') || hasDefinedKey(options, 'file')) && hasDefinedKey(options, 'fileName')) {
    let fileName = options.fileName;
    let md5sum = SparkMD5.ArrayBuffer.hash(options.file);
    let mtime = Date.now();
    let fileSize = options.file.byteLength;
    fileUploadData = {
      fileName,
      md5sum,
      mtime,
      fileSize
    };
    fetchConfig['body'] = "md5=".concat(md5sum, "&filename=").concat(fileName, "&filesize=").concat(fileSize, "&mtime=").concat(mtime);
  }
  if (options.uploadRegisterOnly === true) {
    const {
      fileName,
      fileSize,
      md5sum,
      mtime
    } = options;
    fileUploadData = {
      fileName,
      md5sum,
      mtime,
      fileSize
    };
    fetchConfig['body'] = "md5=".concat(md5sum, "&filename=").concat(fileName, "&filesize=").concat(fileSize, "&mtime=").concat(mtime);
  }

  // checking against access-control-allow-methods seems to be case-sensitive
  fetchConfig.method = fetchConfig.method.toUpperCase();
  fetchConfig.headers = headers;
  options.retryCount = 0;
  if (options.pretend) {
    const response = new PretendResponse({
      url,
      fetchConfig
    }, options, null);
    return _objectSpread$1(_objectSpread$1({
      response
    }, config), {}, {
      source: 'request'
    });
  }
  let rawResponse = await fetch(url, fetchConfig);
  if (isTransientFailure(rawResponse) && options.retry > 0) {
    let retriesCounter = options.retry;
    let nextRetryDelay = typeof options.retryDelay === 'number' ? options.retryDelay : 1;
    while (retriesCounter > 0) {
      await sleep(nextRetryDelay);
      options.retryCount++;
      rawResponse = await fetch(url, fetchConfig);
      if (!isTransientFailure(rawResponse)) {
        break;
      }
      if (typeof options.retryDelay !== 'number') {
        nextRetryDelay *= 2;
      }
      retriesCounter--;
    }
  }
  if ((hasDefinedKey(options, 'file') || hasDefinedKey(options, 'filePatch')) && hasDefinedKey(options, 'fileName') || options.uploadRegisterOnly === true) {
    if (rawResponse.ok) {
      let authData = await rawResponse.json();
      if ('exists' in authData && authData.exists) {
        response = new FileUploadResponse(_objectSpread$1(_objectSpread$1({}, fileUploadData), authData), options, rawResponse);
      } else {
        if (options.uploadRegisterOnly === true) {
          throw new ErrorResponse('API did not recognize provided file meta.', 'Attempted to register existing file, but API did not recognize provided file meta.', rawResponse, options);
        }
        let uploadResponse, isUploadSuccessful, registerResponse;
        if (hasDefinedKey(options, 'filePatch')) {
          const uploadQuery = makeUrlQuery(_objectSpread$1(_objectSpread$1({}, options), {}, {
            upload: authData.uploadKey
          }), filePatchQueryParamNames);
          const uploadUrl = "".concat(options.apiScheme, "://").concat(options.apiAuthorityPart, "/").concat(options.apiPath).concat(path).concat(uploadQuery);
          delete fetchConfig.headers['Content-Type'];
          // upload file patch request
          uploadResponse = await fetch(uploadUrl, _objectSpread$1(_objectSpread$1({}, fetchConfig), {}, {
            method: 'PATCH',
            body: options.filePatch
          }));
          isUploadSuccessful = uploadResponse.status === 204;
        } else {
          let prefix = new Uint8ClampedArray(authData.prefix.split('').map(e => e.charCodeAt(0)));
          let suffix = new Uint8ClampedArray(authData.suffix.split('').map(e => e.charCodeAt(0)));
          let body = new Uint8ClampedArray(prefix.byteLength + options.file.byteLength + suffix.byteLength);
          body.set(prefix, 0);
          body.set(new Uint8ClampedArray(options.file), prefix.byteLength);
          body.set(suffix, prefix.byteLength + options.file.byteLength);

          // full file upload request
          uploadResponse = await fetch(authData.url, {
            headers: {
              [acceptedHeaderNames['contentType']]: authData.contentType
            },
            method: 'POST',
            body: body.buffer
          });
          isUploadSuccessful = uploadResponse.status === 201;

          // register file request
          registerResponse = await fetch(url, _objectSpread$1(_objectSpread$1({}, fetchConfig), {}, {
            body: "upload=".concat(authData.uploadKey)
          }));
          if (!registerResponse.ok) {
            return await throwErrorResponse(registerResponse, options, 'Upload stage 3: ');
          }
        }
        if (isUploadSuccessful) {
          response = new FileUploadResponse(fileUploadData, options, rawResponse, uploadResponse, registerResponse);
        } else {
          return await throwErrorResponse(uploadResponse, options, 'Upload stage 2: ');
        }
      }
    } else {
      return await throwErrorResponse(rawResponse, options, 'Upload stage 1: ');
    }
  } else {
    if (rawResponse.status < 200 || rawResponse.status >= 400) {
      return await throwErrorResponse(rawResponse, options, '');
    }
    let content;
    if (options.format === 'json') {
      try {
        content = await rawResponse.json();
      } catch (_unused) {
        content = null;
      }
      switch (options.method.toUpperCase()) {
        case 'GET':
        default:
          if ('library' in options.resource) {
            if (content && Array.isArray(content)) {
              response = new MultiReadResponse(content, options, rawResponse);
            } else {
              response = new SingleReadResponse(content, options, rawResponse);
            }
          } else {
            if ('schema' in options.resource) {
              response = new SchemaResponse(content, options, rawResponse);
            } else {
              response = new ApiResponse(content, options, rawResponse);
            }
          }
          break;
        case 'POST':
        case 'PUT':
        case 'PATCH':
          if (content && typeof content === 'object' && 'success' in content) {
            response = new MultiWriteResponse(content, options, rawResponse);
          } else {
            response = new SingleWriteResponse(content, options, rawResponse);
          }
          break;
        case 'DELETE':
          response = new DeleteResponse(content, options, rawResponse);
          break;
      }
    } else {
      if ('file' in options.resource && options.method.toUpperCase() === 'GET') {
        let rawData = await rawResponse.arrayBuffer();
        response = new FileDownloadResponse(rawData, options, rawResponse);
      } else if ('fileUrl' in options.resource && options.method.toUpperCase() === 'GET') {
        var _context2;
        const url = await rawResponse.text();
        response = new FileUrlResponse(_trimInstanceProperty(_context2 = url.replace('\n', '')).call(_context2), options, rawResponse);
      } else {
        response = new RawApiResponse(rawResponse, options);
      }
    }
  }
  return _objectSpread$1(_objectSpread$1({}, config), {}, {
    source: 'request',
    response
  });
};

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), true).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }

/**
 * @module zotero-api-client
 */

/**
 * Wrapper function creates closure scope and calls api()
 * @return {Object} Partially configured api functions
 */
const api = function api() {
  /**
   * Entry point of the interface. Configures authentication.
   * Can be used to configure any other properties of the api
   * Returns a set of functions that are bound to that configuration
   * and can be called to specify further api configuration.
   * @param  {String} key  - Authentication key
   * @param  {Object} opts - Optional api configuration. For a list of all
   *                         possible properties, see documentation for
   *                         request() function
   * @return {Object} Partially configured api functions
   * @chainable
   */
  const api = function api() {
    let key = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
    let opts = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    let props = _objectSpread({}, processOpts(opts));
    if (!('executors' in props) && (!this || !('executors' in this))) {
      props.executors = [request];
    }
    if (key) {
      props.zoteroApiKey = key;
    }
    return ef.bind(this)(props);
  };

  /**
   * Configures which library api requests should use.
   * @param {*} [typeOrKey] - Library key, e.g. g1234. Alternatively, if
   *                           the second parameter is present, library type i.e.
   *                          either 'group' or 'user'
   * @param {Number} [id]   - Only when first argument is a type, library id
   * @return {Object} Partially configured api functions
   * @chainable
   */
  const library = function library(typeOrKey) {
    let id = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
    let libraryKey;
    if (arguments.length > 1) {
      switch (typeOrKey.toLowerCase()) {
        case 'user':
          libraryKey = "u".concat(id);
          break;
        case 'group':
          libraryKey = "g".concat(id);
          break;
        default:
          throw new Error("Unrecognized library type \"".concat(typeOrKey, "\""));
      }
    } else {
      libraryKey = typeOrKey;
    }
    return efr.bind(this)({
      library: libraryKey
    });
  };

  /**
   * Configures api to use items or a specific item
   * Can be used in conjunction with library(), collections(), top(), trash(),
   * children(), tags() and any execution function (e.g. get(), post())
   * @param  {String} items - Item key, if present, configure api to point at
   *                          this specific item
   * @return {Object} Partially configured api functions
   * @chainable
   */
  const items = function items() {
    let items = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
    return efr.bind(this)({
      items
    });
  };

  /**
   * Configure api to request all item types
   * Can only be used in conjunction with get()
   * @return {Object} Partially configured api functions
   * @chainable
   */
  const itemTypes = function itemTypes() {
    return efr.bind(this)({
      itemTypes: null
    });
  };

  /**
   * Configure api to request all item fields
   * Can only be used in conjunction with get()
   * @return {Object} Partially configured api functions
   * @chainable
   */
  const itemFields = function itemFields() {
    return efr.bind(this)({
      itemFields: null
    });
  };

  /**
   * Configure api to request localized creator fields
   * Can only be used in conjunction with get()
   * @return {Object} Partially configured api functions
   * @chainable
   */
  const creatorFields = function creatorFields() {
    return efr.bind(this)({
      creatorFields: null
    });
  };

  /**
   * Configure api to request schema
   * Can only be used in conjunction with get()
   * @returns {Object} Partially configured api functions
   * @chainable
   */
  const schema = function schema() {
    return efr.bind(this)({
      schema: null
    });
  };

  /**
   * Configure api to request all valid fields for an item type
   * Can only be used in conjunction with get()
   * @param  {String} itemType - item type for which valid fields will be
   *                             requested, e.g. 'book' or 'journalType'
   * @return {Object} Partially configured api functions
   * @chainable
   */
  const itemTypeFields = function itemTypeFields(itemType) {
    if (!itemType) {
      throw new Error('itemTypeFields() requires an itemType argument');
    }
    return efr.bind(this)({
      itemTypeFields: null
    }, {
      itemType
    });
  };

  /**
   * Configure api to request valid creator types for an item type
   * Can only be used in conjunction with get()
   * @param  {String} itemType - item type for which valid creator types
   *                             will be requested, e.g. 'book' or
   *                             'journalType'
   * @return {Object} Partially configured api functions
   * @chainable
   */
  const itemTypeCreatorTypes = function itemTypeCreatorTypes(itemType) {
    if (!itemType) {
      throw new Error('itemTypeCreatorTypes() requires an itemType argument');
    }
    return efr.bind(this)({
      itemTypeCreatorTypes: null
    }, {
      itemType
    });
  };

  /**
   * Configure api to request template for a new item
   * Can only be used in conjunction with get()
   * @param  {String} itemType - item type for which template will be requested, e.g. 'book' or 'journalType'
   * @param  {String} subType - annotationType if itemType is 'annotation' or linkMode if itemType is 'attachment'
   * @return {Object} Partially configured api functions
   * @chainable
   */
  const template = function template(itemType, subType) {
    if (!itemType) {
      throw new Error('template() requires an itemType argument');
    }
    const subTypeOpts = {};
    if (subType && itemType === 'annotation') {
      subTypeOpts.annotationType = subType;
    } else if (subType && itemType === 'attachment') {
      subTypeOpts.linkMode = subType;
    }
    return efr.bind(this)({
      template: null
    }, _objectSpread({
      itemType
    }, subTypeOpts));
  };

  /**
   * Configure api to use collections or a specific collection
   * Can be used in conjunction with library(), items(), top(), tags() and
   * any of the execution function (e.g. get(), post())
   * @param  {String} collections - Collection key, if present, configure api to
   *                          point to this specific collection
   * @return {Object} Partially configured api functions
   * @chainable
   */
  const collections = function collections(_collections) {
    return efr.bind(this)({
      collections: _collections || null
    });
  };

  /**
   * Configure api to use subcollections that reside underneath the specified
   * collection.
   * Should only be used in conjunction with both library() and collections()
   * and any of the execution function (e.g. get(), post())
   * @return {Object} Partially configured api functions
   * @chainable
   */
  const subcollections = function subcollections() {
    return efr.bind(this)({
      subcollections: null
    });
  };

  /**
   * Configure api to narrow the request to only consider items filed under
   * "My Publications"
   * Should only be used in conjunction with both library() and items()
   * and any of the execution function (e.g. get(), post())
   * @return {Object} Partially configured api functions
   * @chainable
   */
  const publications = function publications() {
    return efr.bind(this)({
      publications: null
    });
  };
  /**
   * Configure api to request or delete tags or request a specific tag
   * Can be used in conjunction with library(), items(), collections() and
   * any of the following execution functions: get(), delete() but only
   * if the first argument is not present. Otherwise, can only be used in
   * conjunction with get()
   * @param  {String} tags - name of a tag to request. If present, configure
   *                         api to request a specific tag.
   * @return {Object} Partially configured api functions
   * @chainable
   */
  const tags = function tags() {
    let tags = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
    return efr.bind(this)({
      tags
    });
  };

  /**
   * Configure api to use saved searches or a specific saved search
   * Can be used in conjunction with library() and any of the execution
   * functions
   * @param  {String} searches - Search key, if present, configure api to point at
   *                             this specific saved search
   * @return {Object} Partially configured api functions
   * @chainable
   */
  const searches = function searches() {
    let searches = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
    return efr.bind(this)({
      searches: searches
    });
  };

  /**
   * Configure api to narrow the request only to the top level items
   * Can be used in conjunction with items() and collections() and only
   * with conjunction with a get() execution function
   * @return {Object} Partially configured api functions
   * @chainable
   */
  const top = function top() {
    return efr.bind(this)({
      top: null
    });
  };

  /**
   * Configure api to narrow the request only to the items in the trash
   * Can be only used in conjunction with items() and get() execution
   * function
   * @return {Object} Partially configured api functions
   * @chainable
   */
  const trash = function trash() {
    return efr.bind(this)({
      trash: null
    });
  };

  /**
   * Configure api to narrow the request only to the children of given
   * item
   * Can be only used in conjunction with items() and get() execution
   * function
   * @return {Object} Partially configured api functions
   * @chainable
   */
  const children = function children() {
    return efr.bind(this)({
      children: null
    });
  };

  /**
   * Configure api to request settings
   * Can only be used in conjunction with get(), put(), post() and delete()
   * For usage with put() and delete() a settings key must be provided
   * For usage with post() a settings key must not be included
   * @param  {String} settings - Settings "key", if present, configures api to point at
   *                             this specific key within settings, e.g. `tagColors`.
  	 * @return {Object} Partially configured api functions
   * @chainable
   */
  const settings = function settings() {
    let settings = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
    return efr.bind(this)({
      settings
    });
  };

  /**
   * Configure api to request deleted content
   * Can only be used in conjunction with get()
   * @param  {Number} since - library version to request deletions since
   * @return {Object} Partially configured api functions
   * @chainable
   */
  const deleted = function deleted(since) {
    const resource = _objectSpread(_objectSpread({}, this.resource), {}, {
      deleted: null
    });
    return ef.bind(this)({
      since,
      resource
    });
  };

  /**
   * Configure api to request user-accessible groups (i.e. The set of groups
   * the current API key has access to, including public groups the key owner
   * belongs to even if the key doesn't have explicit permissions for them.)
   * Can only be used in conjunction with get()
   * @return {Object} Partially configured api functions
   * @chainable
   */
  const groups = function groups() {
    return efr.bind(this)({
      groups: null
    });
  };

  /**
   * Configure api to specify a local version of a given entity.
   * When used in conjunction with the get() exec function, it will populate the
   * If-Modified-Since-Version header.
   * When used in conjunction with post(), put(), patch() or delete(), it will
   * populate the If-Unmodified-Since-Version header.
   * @param  {Number} version - local version of the entity
   * @return {Object} Partially configured api functions
   * @chainable
   */
  const version = function version(_version) {
    if (typeof _version !== 'number' || isNaN(_version)) {
      throw new Error('version() requires a number argument');
    }
    return ef.bind(this)({
      version: _version
    });
  };

  /**
   * Configure api to upload or download an attachment file.
   * Can be only used in conjunction with items() and post()/get()/patch().
   * Method patch() can only be used to upload a binary patch, in this case the last two arguments
   * must be provided.
   * Method post() is used for full uploads. If `md5sum` is provided, it will update an existing
   * file, otherwise it uploads a new file. The last two arguments are not used in this scenario.
   * Method get() is used for downloads, in this case skip all arguments.
   * Use items() to select the attachment item for which the file is uploaded/downloaded.
   * Will populate format on download as well as Content-Type, If*Match headers in case of upload.
   * @param {String} [fileName] - For upload: name of the file, should match values in attachment item entry
   * @param {ArrayBuffer} [file] - New file to be uploaded
   * @param {Number} [mtime] - New file's mtime, leave empty to assume current date/time
   * @param {String} [md5sum] - MD5 hash of an existing file, required for uploads that update existing file
   * @param {ArrayBuffer} [patch] - Binary patch, to be applied to the old file, to produce a new file
   * @param {String} [algorithm] - Algorithm used to compute a diff: xdelta, vcdiff or bsdiff
   * @return {Object} Partially configured api functions
   * @chainable
   */
  const attachment = function attachment(fileName, file, mtime, md5sum, patch, algorithm) {
    let resource = _objectSpread(_objectSpread({}, this.resource), {}, {
      file: null
    });
    let bindParams = {};
    if (md5sum) {
      bindParams.ifMatch = md5sum;
      if (patch && algorithm) {
        bindParams.filePatch = patch;
        bindParams.algorithm = algorithm;
      }
    } else {
      bindParams.ifNoneMatch = '*';
    }
    if (fileName && file) {
      return ef.bind(this)(_objectSpread({
        format: null,
        contentType: 'application/x-www-form-urlencoded',
        fileName,
        resource,
        file,
        mtime
      }, bindParams));
    } else {
      return ef.bind(this)({
        format: null,
        resource
      });
    }
  };

  /**
   * Advanced function that will attempt to register an existing file with a given attachment item
   * based on known file metadata. Can also be used to rename an existing file.
   * Can be only used in conjunction with items() and post().
   * Use items() to select the attachment item for which a file is registered.
   * Will populate Content-Type, If-Match headers.
   * Will fail with a ErrorResponse if API does not return "exists".
   * @param  {String} fileName  - name of the file, should match value in the item, unless renaming
   * @param  {Number} fileSize  - size of the existing file
   * @param  {Number} mtime     - mtime of the existing file
   * @param  {String} md5sum    - md5sum of the existing file
   * @return {Object} Partially configured api functions
   * @chainable
   */
  const registerAttachment = function registerAttachment(fileName, fileSize, mtime, md5sum) {
    let resource = _objectSpread(_objectSpread({}, this.resource), {}, {
      file: null
    });
    if (fileName && typeof fileSize !== 'undefined' && typeof mtime !== 'undefined' && md5sum) {
      return ef.bind(this)({
        contentType: 'application/x-www-form-urlencoded',
        fileName,
        fileSize,
        format: null,
        ifMatch: md5sum,
        md5sum,
        mtime,
        resource,
        uploadRegisterOnly: true
      });
    } else {
      throw new Error('Called registerAttachment() without specifying required parameters');
    }
  };

  /**
   * Configure api to request a temporary attachment file url
   * Can be only used in conjunction with items() and get()
   * Use items() to select the attachment item for which file url is requested
   * Will populate format, redirect.
   * @return {Object} Partially configured api functions
   * @chainable
   */
  const attachmentUrl = function attachmentUrl() {
    let resource = _objectSpread(_objectSpread({}, this.resource), {}, {
      fileUrl: null
    });
    return ef.bind(this)({
      format: null,
      resource
    });
  };

  /**
   * Configure api to request information on the API key.
   * Can only be used in conjunction with get()
   *
   * @return {Object} Partially configured api functions
   * @chainable
   */
  const verifyKeyAccess = function verifyKeyAccess() {
    return efr.bind(this)({
      verifyKeyAccess: null
    });
  };

  /**
   * Execution function. Specifies that the request should use a GET method.
   * @param  {Object} opts - Optional api configuration. If duplicate,
   *                         overrides properties already present. For a list
   *                         of all possible properties, see documentation
   *                         for request() function
   * @return {Promise} A promise that will eventually return either an
   *                   ApiResponse, SingleReadResponse or MultiReadResponse.
   *                   Might throw Error or ErrorResponse.
   */
  const get = function get(opts) {
    return execute(prepareRequest(this, 'get', opts));
  };

  /**
   * Execution function. Specifies that the request should use a POST method.
   * @param  {Array} data  - An array of entities to post
   * @param  {Object} opts - Optional api configuration. If duplicate,
   *                         overrides properties already present. For a list
   *                         of all possible properties, see documentation
   *                         for request() function
   * @return {Promise} A promise that will eventually return MultiWriteResponse.
   *                   Might throw Error or ErrorResponse
   */
  const post = function post(data, opts) {
    return execute(prepareRequest(this, 'post', opts, data));
  };

  /**
   * Execution function. Specifies that the request should use a PUT method.
   * @param  {Object} data - An entity to put
   * @param  {Object} opts - Optional api configuration. If duplicate,
   *                         overrides properties already present. For a list
   *                         of all possible properties, see documentation
   *                         for request() function
   * @return {Promise} A promise that will eventually return SingleWriteResponse.
   *                   Might throw Error or ErrorResponse
   */
  const put = function put(data, opts) {
    return execute(prepareRequest(this, 'put', opts, data));
  };

  /**
   * Execution function. Specifies that the request should use a PATCH
   * method.
   * @param  {Object} data - Partial entity data to patch
   * @param  {Object} opts - Optional api configuration. If duplicate,
   *                         overrides properties already present. For a list
   *                         of all possible properties, see documentation
   *                         for request() function
   * @return {Promise} A promise that will eventually return SingleWriteResponse.
   *                   Might throw Error or ErrorResponse
   */
  const patch = function patch(data, opts) {
    return execute(prepareRequest(this, 'patch', opts, data));
  };

  /**
   * Execution function. Specifies that the request should use a DELETE
   * method.
   * @param  {Array} keysToDelete - An array of keys to delete. Depending on
   *                                how api has been configured, these will
   *                                be item keys, collection keys, search
   *                                keys or tag names. If not present, api
   *                                should be configured to use specific
   *                                item, collection, saved search or settings
   *                                key, in which case, that entity will be deleted
   * @param  {Object} opts - Optional api configuration. If duplicate,
   *                         overrides properties already present. For a list
   *                         of all possible properties, see documentation
   *                         for request() function
   * @return {Promise} A promise that will eventually return DeleteResponse.
   *                   Might throw Error or ErrorResponse
   */
  const del = function del(keysToDelete, opts) {
    return execute(prepareRequest(this, 'delete', opts, keysToDelete));
  };

  /**
   * Execution function. Returns current config without doing any requests.
   * Usually used in advanced scenarios where config needs to be tweaked
   * manually before submitted to the request method or as a debugging tool.
   * @return {Object} current config
   */
  const getConfig = function getConfig() {
    return this;
  };

  /**
   * Execution function. Prepares the request but does not execute fetch(),
   * instead returning a "pretended" response where details for the actual
   * fetch that would have been used are included.
   * Usually used in advanced scenarios where config needs to be tweaked
   * manually before it is submitted to the request method or as a debugging tool.
   * @param  {String} verb - Defines which execution function is used to prepare
   *                         the request. Should be one of 'get', 'post', 'patch'
   *                         'put', 'delete'. Defaults to 'get'.
   * @param  {Object} data - This argument is passed over to the actual execution
   *                         function. For 'get' it is ignored, for 'post', 'patch'
   *                         and 'put' see 'data' of that execution function, for
   *                         'delete' see 'keysToDelete'
   * @param  {Object} opts - Optional api configuration. If duplicate,
   *                         overrides properties already present. For a list
   *                         of all possible properties, see documentation
   *                         for request() function
   * @return {Promise} A promise that will eventually return PretendResponse.
   *                   Might throw Error or ErrorResponse
   */
  const pretend = function pretend() {
    let verb = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'get';
    let data = arguments.length > 1 ? arguments[1] : undefined;
    let opts = arguments.length > 2 ? arguments[2] : undefined;
    return execute(prepareRequest(this, verb, _objectSpread(_objectSpread({}, opts), {}, {
      pretend: true
    }), data));
  };

  /**
   * Used for extending capabilities of the library by installing plugins.
   * In most cases plugins inject additional executors or bind api to an
   * alternative/extended set of functions
   * @param  {function} extend  - function that installs alternative
   *                              or additional functionality of the api.
   *                              It should return bound api functions,
   *                              usually by calling arguments[0].ef()
   * @return {Object} Extended/partially configured api functions
   */
  const use = function use(extend) {
    return extend({
      config: this,
      functions,
      ef,
      efr,
      execute
    });
  };
  const functions = {
    api,
    attachment,
    attachmentUrl,
    children,
    collections,
    creatorFields,
    delete: del,
    //delete is a keyword
    deleted,
    get,
    getConfig,
    groups,
    itemFields,
    items,
    itemTypeCreatorTypes,
    itemTypeFields,
    itemTypes,
    library,
    patch,
    post,
    pretend,
    publications,
    put,
    registerAttachment,
    schema,
    searches,
    settings,
    subcollections,
    tags,
    template,
    top,
    trash,
    use,
    verifyKeyAccess,
    version
  };
  const ef = function ef(opts) {
    let context = _objectSpread(_objectSpread({}, this), opts);
    let enhancedFunctions = {};
    for (let fname in functions) {
      enhancedFunctions[fname] = functions[fname].bind(context);
    }
    return enhancedFunctions;
  };
  const efr = function efr(resource, opts) {
    resource = _objectSpread(_objectSpread({}, this.resource), resource);
    opts = _objectSpread(_objectSpread({}, opts), {}, {
      resource
    });
    return ef.bind(this)(opts);
  };
  const processOpts = function processOpts(opts) {
    let newOpts = _objectSpread({}, opts);
    if ('apiScheme' in newOpts) {
      var _context;
      //  only alphanumeric characters and the +, -, and . allowed
      if (_endsWithInstanceProperty(_context = newOpts.apiScheme).call(_context, '://')) {
        newOpts.apiScheme = newOpts.apiScheme.substring(0, newOpts.apiScheme.length - 3);
      }
      if (!newOpts.apiScheme.match(/^[a-zA-Z0-9+\-.]+$/)) {
        throw new Error('apiScheme can only contain alphanumeric characters, plus (+), minus (-), and dot (.)');
      }
    }
    if ('apiPath' in newOpts) {
      var _context2, _context3;
      if (_startsWithInstanceProperty(_context2 = newOpts.apiPath).call(_context2, '/')) {
        newOpts.apiPath = newOpts.apiPath.substring(1);
      }
      if (newOpts.apiPath.length > 0 && !_endsWithInstanceProperty(_context3 = newOpts.apiPath).call(_context3, '/')) {
        newOpts.apiPath += '/';
      }
    }
    return newOpts;
  };
  const prepareRequest = function prepareRequest(config, verb, opts, body) {
    const method = verb.toLowerCase();
    let relevantSearchKey, requestConfig, keysToDelete;
    switch (method) {
      case 'get':
        requestConfig = _objectSpread(_objectSpread(_objectSpread({}, config), processOpts(opts)), {}, {
          method
        });
        if ('version' in requestConfig) {
          requestConfig['ifModifiedSinceVersion'] = requestConfig['version'];
          delete requestConfig['version'];
        }
        return requestConfig;
      case 'post':
      case 'put':
      case 'patch':
        requestConfig = _objectSpread(_objectSpread(_objectSpread({}, config), opts), {}, {
          body,
          method
        });
        if ('version' in requestConfig) {
          requestConfig['ifUnmodifiedSinceVersion'] = requestConfig['version'];
          delete requestConfig['version'];
        }
        return requestConfig;
      case 'delete':
        requestConfig = _objectSpread(_objectSpread(_objectSpread({}, config), opts), {}, {
          method
        });
        keysToDelete = body;
        if (keysToDelete && !Array.isArray(keysToDelete)) {
          throw new Error("Called delete() with ".concat(typeof keysToDelete, ", expected an Array"));
        }
        if ('version' in requestConfig) {
          requestConfig['ifUnmodifiedSinceVersion'] = requestConfig['version'];
          delete requestConfig['version'];
        }
        if ('resource' in requestConfig && 'items' in requestConfig.resource) {
          relevantSearchKey = 'itemKey';
        } else if ('resource' in requestConfig && 'collections' in requestConfig.resource) {
          relevantSearchKey = 'collectionKey';
        } else if ('resource' in requestConfig && 'tags' in requestConfig.resource) {
          relevantSearchKey = 'tag';
        } else if ('resource' in requestConfig && 'searches' in requestConfig.resource) {
          relevantSearchKey = 'searchKey';
        } else if ('resource' in requestConfig && 'settings' in requestConfig.resource) {
          if (keysToDelete) {
            throw new Error('Arguments to delete() not supported when deleting settings');
          }
        } else {
          throw new Error('Called delete() without first specifying what to delete');
        }
        if (keysToDelete) {
          requestConfig[relevantSearchKey] = [...(requestConfig[relevantSearchKey] || []), ...keysToDelete];
        }
        return requestConfig;
    }
  };
  const execute = async config => {
    for (let f of config.executors) {
      config = await f(config);
    }
    return config.response;
  };
  return api(...arguments);
};

exports.ApiResponse = ApiResponse;
exports.DeleteResponse = DeleteResponse;
exports.ErrorResponse = ErrorResponse;
exports.FileDownloadResponse = FileDownloadResponse;
exports.FileUploadResponse = FileUploadResponse;
exports.FileUrlResponse = FileUrlResponse;
exports.MultiReadResponse = MultiReadResponse;
exports.MultiWriteResponse = MultiWriteResponse;
exports.PretendResponse = PretendResponse;
exports.RawApiResponse = RawApiResponse;
exports.SchemaResponse = SchemaResponse;
exports.SingleReadResponse = SingleReadResponse;
exports.SingleWriteResponse = SingleWriteResponse;
exports.default = api;
exports.request = request;
